larpandoracontent/LArTwoDReco/HitWidthClusterMergingAlgorithm.cc

```diff
@@ -0,0 +1,503 @@
+            throw StatusCodeException(STATUS_CODE_NOT_FOUND);
+
+        for (ClusterVector::const_iterator iterTestCluster = iterCurrentCluster; iterTestCluster != clusterVector.end(); ++iterTestCluster)
+        {
+            if (iterCurrentCluster == iterTestCluster)
+                continue;
+
+	    const Cluster *const pTestCluster = *iterTestCluster;
+            const LArHitWidthHelper::ClusterToParametersMap::const_iterator testParametersIter = pClusterToParametersMap->find(pTestCluster);
+
+	    if(testParametersIter == pClusterToParametersMap->end())
+                throw StatusCodeException(STATUS_CODE_NOT_FOUND);
+
+            if (!this->AreClustersAssociated(currentParametersIter->second, testParametersIter->second))
+	        continue;
+
+            clusterAssociationMap[pCurrentCluster].m_forwardAssociations.insert(pTestCluster);
+            clusterAssociationMap[pTestCluster].m_backwardAssociations.insert(pCurrentCluster);
+        }
+    }
+
+    // remove all 'shortcut' routes
+    this->CleanupClusterAssociations(clusterVector, clusterAssociationMap);
+
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void HitWidthClusterMergingAlgorithm::CleanupClusterAssociations(const ClusterVector &clusterVector, ClusterAssociationMap &clusterAssociationMap) const
+{
+
+  // Create temporary map so can delete elements whilst still iterating over them
+  ClusterAssociationMap tempMap(clusterAssociationMap);
+
+  for(const Cluster *const pCluster : clusterVector) 
+  {
+      const ClusterAssociationMap::const_iterator primaryMapIter = clusterAssociationMap.find(pCluster);
+
+      if(primaryMapIter == clusterAssociationMap.end())
+          continue;
+
+      const ClusterSet &primaryForwardAssociations(primaryMapIter->second.m_forwardAssociations);
+
+      // loop through the primary associations of each cluster
+      // remove clusters that are present in secondary associations of other primary clusters 
+      for(const Cluster *const pConsideredCluster : primaryForwardAssociations)
+      {
+          for(const Cluster *const pPrimaryCluster : primaryForwardAssociations)
+          {
+              if(pConsideredCluster == pPrimaryCluster)
+                  continue;
+
+              const ClusterAssociationMap::const_iterator secondaryMapIter = clusterAssociationMap.find(pPrimaryCluster);
+
+	          // if primary cluster has no associations (this shouldn't ever be the case)
+              if(secondaryMapIter == clusterAssociationMap.end()) 
+                  continue;
+
+              const ClusterSet &secondaryForwardAssociations(secondaryMapIter->second.m_forwardAssociations);
+
+              // if cluster is present in the forward associations of any cluster at the same level remove
+              if(secondaryForwardAssociations.find(pConsideredCluster) != secondaryForwardAssociations.end()) 
```

Comment:
Absolutely fine to look up what's in the ClusterSet here; just looping over this container type that we should avoid.